Create cluster: panic: runtime error with eksctl version 0.1.39

**What happened?**
Panic: runtime error: invalid memory address or nil pointer dereference

**What you expected to happen?**
After switching from 0.1.38 to 0.1.39 `eksctl create cluster -f cluster.yaml` is broken for me

**How to reproduce it?**
Used config (cluster.yaml):
```
 apiVersion: eksctl.io/v1alpha5
 kind: ClusterConfig

 metadata:
   name: dev-eks
   region: eu-central-1

 vpc:
   subnets:
     private:
       eu-central-1a: {id: subnet-00000000000000000}
       eu-central-1b: {id: subnet-00000000000000001}
       eu-central-1c: {id: subnet-00000000000000002}

 nodeGroups:
   - name: ng-1a
     ami: ami-0b7127e7a2a38802a
     instanceType: t3.medium
     desiredCapacity: 1
     availabilityZones: ["eu-central-1a"]
     privateNetworking: true
     ssh:
       publicKeyName: bla.bla
     iam:
       attachPolicyARNs:
         - arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy
         - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
         - arn:aws:iam::000000000000:policy/ManageRoute53HostedZone
```
**Versions**
Please paste in the output of these commands:
```
$ eksctl version
[ℹ]  version.Info{BuiltAt:"", GitCommit:"", GitTag:"0.1.39"}
```

**Logs**
```
eksctl create cluster -f cluster.yaml
[ℹ]  using region eu-central-1
[✔]  using existing VPC (vpc-xxx) and subnets (private:[subnet-0... subnet-1... subnet-2...] public:[])
[!]  custom VPC/subnets will be used; if resulting cluster doesn't function as expected, make sure to review the configuration of VPC/subnets
[ℹ]  nodegroup "ng-1a" will use "ami-0b7127e7a2a38802a" [AmazonLinux2/1.13]
[ℹ]  using EC2 key pair "bla.bla"
[ℹ]  using Kubernetes version 1.13
[ℹ]  creating EKS cluster "dev-eks" in "eu-central-1" region
[ℹ]  1 nodegroup (ng-1a) was included
[ℹ]  will create a CloudFormation stack for cluster itself and 1 nodegroup stack(s)
[ℹ]  if you encounter any issues, check CloudFormation console or try 'eksctl utils describe-stacks --region=eu-central-1 --name=dev-eks'
[ℹ]  2 sequential tasks: { create cluster control plane "dev-eks", create nodegroup "ng-1a" }
[ℹ]  building cluster stack "eksctl-dev-eks-cluster"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x20ac393]

goroutine 51 [running]:
github.com/weaveworks/eksctl/pkg/cfn/builder.(*ClusterResourceSet).addOutputsForVPC(0xc00011a280)
	/src/pkg/cfn/builder/vpc.go:160 +0x103
github.com/weaveworks/eksctl/pkg/cfn/builder.(*ClusterResourceSet).AddAllResources(0xc00011a280, 0xc00011a280, 0xc000550e80)
	/src/pkg/cfn/builder/cluster.go:49 +0xac
github.com/weaveworks/eksctl/pkg/cfn/manager.(*StackCollection).createClusterTask(0xc000354f60, 0xc0000407e0, 0x60, 0x0)
	/src/pkg/cfn/manager/cluster.go:33 +0x329
github.com/weaveworks/eksctl/pkg/cfn/manager.(*taskWithoutParams).Do(0xc00010f100, 0xc0000407e0, 0xc0000407e0, 0x1)
	/src/pkg/cfn/manager/tasks.go:132 +0x34
github.com/weaveworks/eksctl/pkg/cfn/manager.doSingleTask(0xc000040780, 0x3a86800, 0xc00010f100, 0x0)
	/src/pkg/cfn/manager/tasks.go:185 +0xec
github.com/weaveworks/eksctl/pkg/cfn/manager.doSequentialTasks(0xc000040780, 0xc00010f160, 0x2, 0x2)
	/src/pkg/cfn/manager/tasks.go:215 +0x70
created by github.com/weaveworks/eksctl/pkg/cfn/manager.(*TaskTree).DoAllSync
	/src/pkg/cfn/manager/tasks.go:115 +0x246
```

--- a/pkg/ctl/cmdutils/configfile.go
+++ b/pkg/ctl/cmdutils/configfile.go
@@ -56,7 +56,9 @@ func (l *commonClusterConfigLoader) Load() error {
 		return l.validateWithoutConfigFile()
 	}
 
-	if err := eks.LoadConfigFromFile(l.ClusterConfigFile, l.ClusterConfig); err != nil {
+	var err error
+
+	if l.ClusterConfig, err = eks.LoadConfigFromFile(l.ClusterConfigFile); err != nil {
 		return err
 	}
 	meta := l.ClusterConfig.Metadata
@@ -150,6 +152,14 @@ func NewCreateClusterLoader(rc *ResourceCmd, ngFilter *NodeGroupFilter) ClusterC
 	l.validateWithConfigFile = func() error {
 		if l.ClusterConfig.VPC == nil {
 			l.ClusterConfig.VPC = api.NewClusterVPC()
+		}
+
+		if l.ClusterConfig.VPC.NAT == nil {
+			l.ClusterConfig.VPC.NAT = api.DefaultClusterNAT()
+		}
+
+		if !api.IsSetAndNonEmptyString(l.ClusterConfig.VPC.NAT.Gateway) {
+			*l.ClusterConfig.VPC.NAT.Gateway = api.ClusterSingleNAT
 		}
 
 		if l.ClusterConfig.HasAnySubnets() && len(l.ClusterConfig.AvailabilityZones) != 0 {

--- a/pkg/ctl/cmdutils/configfile.go
+++ b/pkg/ctl/cmdutils/configfile.go
@@ -58,6 +58,9 @@ func (l *commonClusterConfigLoader) Load() error {
 
 	var err error
 
+	// The reference to ResourceCmd.ClusterConfig should only be reassigned if ClusterConfigFile is specified
+	// because other parts of the code store the pointer locally and access it directly instead of via
+	// the ResourceCmd reference
 	if l.ClusterConfig, err = eks.LoadConfigFromFile(l.ClusterConfigFile); err != nil {
 		return err
 	}

--- a/pkg/ctl/cmdutils/configfile_test.go
+++ b/pkg/ctl/cmdutils/configfile_test.go
@@ -99,5 +99,35 @@ var _ = Describe("cmdutils configfile", func() {
 				Expect(cfg.Metadata.Version).To(BeEmpty())
 			}
 		})
+
+		It("should set VPC.NAT.Gateway with the correct value", func() {
+			natTests := []struct {
+				configFile      string
+				expectedGateway string
+			}{
+				// No VPC set
+				{"01-simple-cluster.yaml", api.ClusterSingleNAT},
+				// VPC set but not NAT
+				{"02-custom-vpc-cidr-no-nodes.yaml", api.ClusterSingleNAT},
+				// VPC and subnets set but not NAT
+				{"04-existing-vpc.yaml", api.ClusterSingleNAT},
+				// NAT set
+				{"09-nat-gateways.yaml", api.ClusterHighlyAvailableNAT},
+			}
+
+			for _, natTest := range natTests {
+				rc := &ResourceCmd{
+					Command:           newCmd(),
+					ClusterConfigFile: filepath.Join(examplesDir, natTest.configFile),
+					ClusterConfig:     api.NewClusterConfig(),
+					ProviderConfig:    &api.ProviderConfig{},
+				}
+
+				Expect(NewCreateClusterLoader(rc, nil).Load()).To(Succeed())
+				cfg := rc.ClusterConfig
+				Expect(cfg.VPC.NAT.Gateway).To(Not(BeNil()))
+				Expect(*cfg.VPC.NAT.Gateway).To(Equal(natTest.expectedGateway))
+			}
+		})
 	})
 })

--- a/pkg/ctl/cmdutils/configfile_test.go
+++ b/pkg/ctl/cmdutils/configfile_test.go
@@ -82,17 +82,16 @@ var _ = Describe("cmdutils configfile", func() {
 
 			Expect(examples).To(HaveLen(10))
 			for _, example := range examples {
-				cfg := api.NewClusterConfig()
-
 				rc := &ResourceCmd{
 					Command:           newCmd(),
 					ClusterConfigFile: example,
-					ClusterConfig:     cfg,
+					ClusterConfig:     api.NewClusterConfig(),
 					ProviderConfig:    &api.ProviderConfig{},
 				}
 
 				err := NewMetadataLoader(rc).Load()
 
+				cfg := rc.ClusterConfig
 				Expect(err).ToNot(HaveOccurred())
 				Expect(cfg.Metadata.Name).ToNot(BeEmpty())
 				Expect(cfg.Metadata.Region).ToNot(BeEmpty())

--- a/pkg/eks/api.go
+++ b/pkg/eks/api.go
@@ -2,10 +2,11 @@ package eks
 
 import (
 	"fmt"
-	"github.com/weaveworks/eksctl/pkg/utils"
 	"io/ioutil"
 	"os"
 	"time"
+
+	"github.com/weaveworks/eksctl/pkg/utils"
 
 	"github.com/kris-nova/logger"
 	"github.com/pkg/errors"
@@ -163,10 +164,11 @@ func New(spec *api.ProviderConfig, clusterSpec *api.ClusterConfig) *ClusterProvi
 }
 
 // LoadConfigFromFile populates cfg based on contents of configFile
-func LoadConfigFromFile(configFile string, cfg *api.ClusterConfig) error {
+func LoadConfigFromFile(configFile string) (*api.ClusterConfig, error) {
 	data, err := readConfig(configFile)
+	// LoadConfigFromFile loads ClusterConfig from configFile
 	if err != nil {
-		return errors.Wrapf(err, "reading config file %q", configFile)
+		return nil, errors.Wrapf(err, "reading config file %q", configFile)
 	}
 
 	// strict mode is not available in runtime.Decode, so we use the parser
@@ -175,22 +177,19 @@ func LoadConfigFromFile(configFile string, cfg *api.ClusterConfig) error {
 	// NOTE: we must use sigs.k8s.io/yaml, as it behaves differently from
 	// github.com/ghodss/yaml, which didn't handle nested structs well
 	if err := yaml.UnmarshalStrict(data, &api.ClusterConfig{}); err != nil {
-		return errors.Wrapf(err, "loading config file %q", configFile)
+		return nil, errors.Wrapf(err, "loading config file %q", configFile)
 	}
 
 	obj, err := runtime.Decode(scheme.Codecs.UniversalDeserializer(), data)
 	if err != nil {
-		return errors.Wrapf(err, "loading config file %q", configFile)
+		return nil, errors.Wrapf(err, "loading config file %q", configFile)
 	}
 
-	cfgLoaded, ok := obj.(*api.ClusterConfig)
+	cfg, ok := obj.(*api.ClusterConfig)
 	if !ok {
-		return fmt.Errorf("decoded object of wrong type")
+		return nil, fmt.Errorf("expected to decode object of type %T; got %T", &api.ClusterConfig{}, cfg)
 	}
-
-	*cfg = *cfgLoaded // mutate the content, not the reference
-
-	return nil
+	return cfg, nil
 }
 
 func readConfig(configFile string) ([]byte, error) {

--- a/pkg/eks/api.go
+++ b/pkg/eks/api.go
@@ -163,10 +163,9 @@ func New(spec *api.ProviderConfig, clusterSpec *api.ClusterConfig) *ClusterProvi
 	return c
 }
 
-// LoadConfigFromFile populates cfg based on contents of configFile
+// LoadConfigFromFile loads ClusterConfig from configFile
 func LoadConfigFromFile(configFile string) (*api.ClusterConfig, error) {
 	data, err := readConfig(configFile)
-	// LoadConfigFromFile loads ClusterConfig from configFile
 	if err != nil {
 		return nil, errors.Wrapf(err, "reading config file %q", configFile)
 	}

--- a/pkg/eks/api_test.go
+++ b/pkg/eks/api_test.go
@@ -17,61 +17,58 @@ import (
 var _ = Describe("eksctl API", func() {
 
 	Context("loading config files", func() {
-		var (
-			cfg *api.ClusterConfig
-		)
+
 		BeforeEach(func() {
 			err := api.Register()
 			Expect(err).ToNot(HaveOccurred())
-			cfg = &api.ClusterConfig{}
 		})
 
 		It("should load a valid YAML config without error", func() {
-			err := LoadConfigFromFile("../../examples/01-simple-cluster.yaml", cfg)
+			cfg, err := LoadConfigFromFile("../../examples/01-simple-cluster.yaml")
 			Expect(err).ToNot(HaveOccurred())
 			Expect(cfg.Metadata.Name).To(Equal("cluster-1"))
 			Expect(cfg.NodeGroups).To(HaveLen(1))
 		})
 
 		It("should load a valid JSON config without error", func() {
-			err := LoadConfigFromFile("testdata/example.json", cfg)
+			cfg, err := LoadConfigFromFile("testdata/example.json")
 			Expect(err).ToNot(HaveOccurred())
 			Expect(cfg.Metadata.Name).To(Equal("cluster-1"))
 			Expect(cfg.NodeGroups).To(HaveLen(1))
 		})
 
 		It("should error when version is a float, not a string", func() {
-			err := LoadConfigFromFile("testdata/bad-type-1.yaml", cfg)
+			_, err := LoadConfigFromFile("testdata/bad-type-1.yaml")
 			Expect(err).To(HaveOccurred())
 			Expect(err.Error()).To(HavePrefix(`loading config file "testdata/bad-type-1.yaml": v1alpha5.ClusterConfig.Metadata: v1alpha5.ClusterMeta.Version: ReadString: expects " or n, but found 1`))
 		})
 
 		It("should reject unknown field in a YAML config", func() {
-			err := LoadConfigFromFile("testdata/bad-field-1.yaml", cfg)
+			_, err := LoadConfigFromFile("testdata/bad-field-1.yaml")
 			Expect(err).To(HaveOccurred())
 			Expect(err.Error()).To(HavePrefix(`loading config file "testdata/bad-field-1.yaml": error unmarshaling JSON: while decoding JSON: json: unknown field "zone"`))
 		})
 
 		It("should reject unknown field in a YAML config", func() {
-			err := LoadConfigFromFile("testdata/bad-field-2.yaml", cfg)
+			_, err := LoadConfigFromFile("testdata/bad-field-2.yaml")
 			Expect(err).To(HaveOccurred())
 			Expect(err.Error()).To(HavePrefix(`loading config file "testdata/bad-field-2.yaml": error unmarshaling JSON: while decoding JSON: json: unknown field "bar"`))
 		})
 
 		It("should reject unknown field in a JSON config", func() {
-			err := LoadConfigFromFile("testdata/bad-field-1.json", cfg)
+			_, err := LoadConfigFromFile("testdata/bad-field-1.json")
 			Expect(err).To(HaveOccurred())
 			Expect(err.Error()).To(HavePrefix(`loading config file "testdata/bad-field-1.json": error unmarshaling JSON: while decoding JSON: json: unknown field "nodes"`))
 		})
 
 		It("should reject old API version", func() {
-			err := LoadConfigFromFile("testdata/old-version.json", cfg)
+			_, err := LoadConfigFromFile("testdata/old-version.json")
 			Expect(err).To(HaveOccurred())
 			Expect(err.Error()).To(HavePrefix(`loading config file "testdata/old-version.json": no kind "ClusterConfig" is registered for version "eksctl.io/v1alpha3" in scheme`))
 		})
 
 		It("should error when cannot read a file", func() {
-			err := LoadConfigFromFile("../../examples/nothing.xml", cfg)
+			_, err := LoadConfigFromFile("../../examples/nothing.xml")
 			Expect(err).To(HaveOccurred())
 			Expect(err.Error()).To(Equal(`reading config file "../../examples/nothing.xml": open ../../examples/nothing.xml: no such file or directory`))
 		})

